[Profiler] Define pattern matcher structure

davidchencsl opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81056
* #80572
* __->__ #80108

Differential Revision: [D37535663](https://our.internmc.facebook.com/intern/diff/D37535663)

Summary:
I define a base Pattern and some helper function in this commit

Test Plan:
Added test in test_profiler.py

davidchencsl committed:
Define pattern matcher structure

[ghstack-poisoned]

davidchencsl committed:
Update on "Define pattern matcher structure"




[ghstack-poisoned]

davidchencsl committed:
Update on "Define pattern matcher structure"




[ghstack-poisoned]

davidchencsl commented:
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37535663).

robieta commented on torch/profiler/_pattern_matcher.py:
```diff
@@ -0,0 +1,88 @@
+        return re.search(self.name, event.name()) is not None
+
+
+def EventTreeDFS(event_tree):
```
I really like that you factored the traversal algorithm into a separate function. Any reason why it's not camel case?

robieta commented on torch/profiler/pattern_matcher.py:
```diff
@@ -1,103 +0,0 @@
-from torch.autograd.profiler import profile
```
Curious that this shows up...

davidchencsl replied:
I drop an accidental commit, seems to be related.

robieta commented on test/test_profiler.py:
```diff
@@ -1282,6 +1282,45 @@ def test_utils_compute_queue_depth_when_no_cuda_events(self):
         basic_evaluation = _utils.BasicEvaluation(prof.profiler)
         self.assertFalse(basic_evaluation.compute_queue_depth())
 
+    @unittest.skipIf(not torch.cuda.is_available(), "CUDA is required")
```
Since we're only testing the ops (CPU side) it's probably to drop cuda so we can test more broadly.

robieta commented on test/test_profiler.py:
```diff
@@ -1282,6 +1282,45 @@ def test_utils_compute_queue_depth_when_no_cuda_events(self):
+        self.assertEqual(None, pattern.next_of(event_tree[-1]))
+        self.assertEqual(event_tree[1], pattern.next_of(event_tree[0]))
+        self.assertEqual(event_tree[0], pattern.prev_of(event_tree[1]))
+
```
nit: extra spaces

robieta reviewed (commented):
I have some minor feedback, but overall looks great.

davidchencsl commented on torch/profiler/_pattern_matcher.py:
```diff
@@ -0,0 +1,88 @@
+
+    def next_of(self, event):
+        siblings = self.siblings_of(event)
+        index = siblings.index(event)
```
💯

davidchencsl commented on torch/profiler/_pattern_matcher.py:
```diff
@@ -0,0 +1,88 @@
+            event = event.parent
+        return event
+
+    def siblings_of(self, event):
```
So the list of events doesn't necessarily guarente the order of execution? Then how can I know the next executed ops (not children)

davidchencsl commented on torch/profiler/_pattern_matcher.py:
```diff
@@ -0,0 +1,88 @@
+        )
+
+    def report(self, event):
+        msg = f"{event.name()}\n{self.description}\n{source_code_location(event)}"
```
I left this here for debugging purpose, I think I will format this to a seperate report class

davidchencsl replied:
Oh I see what you are saying. I would get rid of the event.name()

davidchencsl committed:
Update on "Define pattern matcher structure"


Differential Revision: [D37535663](https://our.internmc.facebook.com/intern/diff/D37535663)

Summary:
I define a base Pattern and some helper function in this commit

Test Plan:
Added test in test_profiler.py

[ghstack-poisoned]

davidchencsl commented:
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37535663).

davidchencsl committed:
Update on "Define pattern matcher structure"


Differential Revision: [D37535663](https://our.internmc.facebook.com/intern/diff/D37535663)

Summary:
I define a base Pattern and some helper function in this commit

Test Plan:
Added test in test_profiler.py

[ghstack-poisoned]

davidchencsl commented:
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37535663).

robieta reviewed (approved):
LGTM

davidchencsl commented:
@pytorchbot merge

davidchencsl committed:
Update on "[Profiler] Define pattern matcher structure"


Differential Revision: [D37535663](https://our.internmc.facebook.com/intern/diff/D37535663)

Summary:
I define a base Pattern and some helper function in this commit

Test Plan:
Added test in test_profiler.py

[ghstack-poisoned]

davidchencsl commented:
@pytorchmergebot merge -g
